### ISSUE TITLE
More intelligent behaviour of code and code-block markdown

### ISSUE THREAD

**Issue by ultra-byte** [laurent22/joplin#1606]
I thought it would be tidier to lump these three minor things into a single issue as they are all interlinked:

### Operating system
- Windows

### Application
- Desktop (1.0.158)

I assume the behaviour is consistent on other platforms but I cannot test that.

---
## 1 - </> button behaviour
Current behaviour of the </> code button is to always use a single ` regardless of how many lines of text are selected. It would be great if:
* when a user has selected two or more lines of text
* and uses the </> button

the editor would instead use code block (```) around the selected text.

**Current behaviour:**
![image](https://user-images.githubusercontent.com/50975926/58528537-fd372e00-822a-11e9-9826-ea77906e9061.png)

**Proposed behaviour:**
![image](https://user-images.githubusercontent.com/50975926/58528565-150eb200-822b-11e9-8c7e-3e0989ea5c3d.png)

---
## 2 - Possible bugs
While experimenting I found some odd cases which seem like possible bugs. Examples 1 & 2 to me are expected behaviour but examples 3 & 4 are unexpected. To elaborate:

* Example 3 - the first line of the code block disappears from the rendered note.
* Example 4 - The terminating ``` don't work and so the code block remains unclosed.

![image](https://user-images.githubusercontent.com/50975926/58530180-e0eabf80-8231-11e9-82ce-9b43ba7e8e72.png)

I'm not sure what the authority for markup is but, Example 3's behaviour is contrary to examples given here:
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code

---
## 3 - Improving auto-complete
When simply trying to type a code block the auto-complete could be a little more helpful. Auto-complete seems to assume you are always trying to type a single line of code where:

* hitting \` once gives: \`|\` (where | is the location of the cursor).
* But hitting \` three times gives you: \```|` where I would expect the behaviour to be: \```|``` .

**Comment by CalebJohn** [comment]
I'll respond in the same order you layed these out. Normally (and this goes for almost all open source projects) we'd request that you keep issues one problem per issue. But in your case only one of these issues can't be addressed.

# 1
This is a good feature request.

# 2
You're misunderstanding the markdown syntax. On the first line of a code block (after the three backticks) you can specify the language of the code.  If you look at the example you linked the only thing on that first line is programming language name, and in the example it even doesn't show up in the rendering so I don't know why you thought this syntax was correct. 
For example 4 this is also part of the code block definition. As per markdown syntax the closing tags need to be on a newline.

In short you're misunderstanding markdown syntax and joplin is showing the correct rendering.

# 3
This was literally the last reported issue #1604 and a fix is already in the pipeline #1599 

Thanks for the report! But in the future please do a quick search to see if your issue is unique.

**Comment by tessus** [comment]
There was a feature request once for a separate icon for code blocks. But I like the idea to make it dependent on the number of lines.

So if the selection includes at least one line break `\n`, a code block is generated.
Not sure how easy that would be though, because what would happen, if you selected text in the middle of a paragraph?

**Comment by ultra-byte** [comment]
Thanks for your quick responses.

## 1
Apologies for not breaking out the issue into separate tickets, I was worried about being spammy and I think I just over thought it (I've also never done this before). Thank you for your understanding.

## 2
Ah I understand now. I'm not sure how I missed that, but yes that makes perfect sense.

## 3
Again sorry for I did try a search and did not find anything. It seems that those two were posted while I took over a day to compose and submit this ticket.

Again thank you, I am absolutely loving Joplin.

**Comment by tessus** [comment]
Just as an info. The [forum](https://discourse.joplinapp.org/) is a good place to start discussions or ask for support.
The github tracker should only be used for (confirmed) bugs and (accepted) feature requets.
